OSGi metadata doesn't export package versions

JNA version: 5.0.0-SNAPSHOT and below!

Hello. It seems that the OSGi metadata for the jna jar files doesn't export package versions. For example:

```
$ bnd print ~/.m2/repository/net/java/dev/jna/jna/4.5.1/jna-4.5.1.jar
[MANIFEST jna-4.5.1]
Ant-Version                              Apache Ant 1.9.9                        
Bundle-ActivationPolicy                  lazy                                    
Bundle-Category                          jni                                     
Bundle-Description                       JNA Library                             
Bundle-ManifestVersion                   2                                       
Bundle-Name                              jna                                     
Bundle-NativeCode                         com/sun/jna/win32-x86/jnidispatch.dll; processor=x86;osname=win32, com/sun/jna/win32-x86-64/jnidispatch.dll; processor=x86-64;osname=win32, com/sun/jna/win32-x86/jnidispatch.dll; processor=x86;osname=win, com/sun/jna/win32-x86-64/jnidispatch.dll; processor=x86-64;osname=win, com/sun/jna/w32ce-arm/jnidispatch.dll; processor=arm;osname=wince,  com/sun/jna/sunos-x86/libjnidispatch.so; processor=x86;osname=sunos, com/sun/jna/sunos-x86-64/libjnidispatch.so; processor=x86-64;osname=sunos, com/sun/jna/sunos-sparc/libjnidispatch.so; processor=sparc;osname=sunos, com/sun/jna/sunos-sparcv9/libjnidispatch.so; processor=sparcv9;osname=sunos,  com/sun/jna/aix-ppc/libjnidispatch.a; processor=ppc;osname=aix, com/sun/jna/aix-ppc64/libjnidispatch.a; processor=ppc64;osname=aix,  com/sun/jna/linux-ppc/libjnidispatch.so; processor=ppc;osname=linux, com/sun/jna/linux-ppc64/libjnidispatch.so; processor=ppc64;osname=linux, com/sun/jna/linux-ppc64le/libjnidispatch.so; processor=ppc64le;osname=linux, com/sun/jna/linux-x86/libjnidispatch.so; processor=x86;osname=linux, com/sun/jna/linux-x86-64/libjnidispatch.so; processor=x86-64;osname=linux, com/sun/jna/linux-arm/libjnidispatch.so; processor=arm;osname=linux, com/sun/jna/linux-armel/libjnidispatch.so; processor=armel;osname=linux, com/sun/jna/linux-aarch64/libjnidispatch.so; processor=aarch64;osname=linux, com/sun/jna/linux-ia64/libjnidispatch.so; processor=ia64;osname=linux, com/sun/jna/linux-sparcv9/libjnidispatch.so; processor=sparcv9;osname=linux, com/sun/jna/linux-mips64el/libjnidispatch.so; processor=mips64el;osname=linux, com/sun/jna/linux-s390x/libjnidispatch.so; processor=S390x;osname=linux,  com/sun/jna/freebsd-x86/libjnidispatch.so; processor=x86;osname=freebsd, com/sun/jna/freebsd-x86-64/libjnidispatch.so; processor=x86-64;osname=freebsd, com/sun/jna/openbsd-x86/libjnidispatch.so; processor=x86;osname=openbsd, com/sun/jna/openbsd-x86-64/libjnidispatch.so; processor=x86-64;osname=openbsd,  com/sun/jna/darwin/libjnidispatch.jnilib; osname=macosx;processor=x86;processor=x86-64;processor=ppc 
Bundle-RequiredExecutionEnvironment      J2SE-1.4                                
Bundle-SymbolicName                      com.sun.jna                             
Bundle-Vendor                            JNA Development Team                    
Bundle-Version                           4.5.1                                   
Created-By                               1.8.0_151-b12 (Oracle Corporation)      
Export-Package                           com.sun.jna,com.sun.jna.ptr,com.sun.jna.win32
Implementation-Title                     com.sun.jna                             
Implementation-Vendor                    JNA Development Team                    
Implementation-Version                   4.5.1 (b0)                              
Main-Class                               com.sun.jna.Native                      
Manifest-Version                         1.0                                     
Specification-Title                      Java Native Access (JNA)                
Specification-Vendor                     JNA Development Team                    
Specification-Version                    4                                       

[IMPEXP]
Export-Package
  com.sun.jna                            
  com.sun.jna.ptr                        
  com.sun.jna.win32                      
```

An example of an OSGi bundle that does export package versions:

```
$ bnd print ~/.m2/repository/com/io7m/junreachable/com.io7m.junreachable.core/2.1.1/com.io7m.junreachable.core-2.1.1.jar
[MANIFEST com.io7m.junreachable.core-2.1.1]
Bundle-Description                       Exception types for marking unreachable/unimplemented code (Core)
Bundle-DocURL                            http://io7m.github.io/junreachable/     
Bundle-License                           http://io7m.com/license/isc.txt         
Bundle-ManifestVersion                   2                                       
Bundle-Name                              com.io7m.junreachable.core              
Bundle-SymbolicName                      com.io7m.junreachable.core              
Bundle-Version                           2.1.1                                   
Export-Package                           com.io7m.junreachable;version="2.1.1"                                 
Manifest-Version                         1.0                                     
Require-Capability                       osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=9.0))"                          

[IMPEXP]
Export-Package
  com.io7m.junreachable                  {version=2.1.1}
```

Note the `com.io7m.junreachable;version="2.1.1"` declaration in `Export-Package`.

The reason that this is an issue is that there's no way to specify that your package requires, for example, version `4.5.1` of  the `jna` packages. You can only specify that you want `jna`, meaning that if the user loads a version you don't support, they'll get `ClassNotFoundException` errors and other binary-incompatibility related problems.

The correct way to fix this is to put version declarations on the exported packages.

I'm having a bit of trouble. No matter which compiler I use (OpenJDK 8, 9, 10), I always get the following error:

```
[rm@almond jna]$ JAVA_HOME=/usr/lib/jvm/java-8-openjdk ant
Buildfile: /home/rm/git/com.github/io7m/jna/build.xml

-enable-native:

-prepare-anttools:

-dynamic-properties:
     [echo] Java version 1.8.0_172, compatibility: 1.6, ant: 1.8
     [echo] JNA version 5.0.0-SNAPSHOT, native 5.2.1, android 5000000
     [echo] OpenJDK 64-Bit Server VM (Oracle Corporation, 25.172-b11)
     [echo] java.home=/usr/lib/jvm/java-8-openjdk/jre
     [echo] java.library.path=/usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib
     [echo] os.prefix=linux-x86-64
     [echo] os.name=Linux
     [echo] os.arch=amd64 (little)
     [echo] build=build
     [echo] build.native=/home/rm/git/com.github/io7m/jna/build/native-linux-x86-64
     [echo] build.headers=/home/rm/git/com.github/io7m/jna/build/headers
     [echo] build.aar=/home/rm/git/com.github/io7m/jna/build/aar

-setup:

compile:

BUILD FAILED
/home/rm/git/com.github/io7m/jna/build.xml:362: javac doesn't support the "nativeheaderdir" attribute

Total time: 0 seconds
```

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -7,6 +7,7 @@ Release 5.0.0 (Next release)
 
 Features
 --------
+* [#975](https://github.com/java-native-access/jna/pull/982): Add package versions to OSGi metadata. - [@io7m](https://github.com/io7m).
 * [#915](https://github.com/java-native-access/jna/pull/915): Adding interfaces to call to Cryptui and Crypt32 windows libraries and adding related structures to Wincrypt. - [@rosh89](https://github.com/rosh89).
 * [#903](https://github.com/java-native-access/jna/pull/903): Carry `HRESULT` in `c.s.j.p.win32.COM.COMException`, introduce `c.s.j.p.win32.COM.COMInvokeException` as subclass of `COMException` for exception as the result of a `IDispatch#Invoke`. The `EXECPINFO` is unwrapped into fields in the `COMInvokeException` and correctly freed. - [@matthiasblaesing](https://github.com/matthiasblaesing).
 * [#822](https://github.com/java-native-access/jna/issues/822): `Native#loadLibrary` requires that the interface class passed in is an instance of Library. The runtime check can be enhanced by using a constraint generic. This breaks binary compatibility (see notes below) - [@d-noll](https://github.com/d-noll).<br /><br />In a followup, the original `loadLibrary` methods were deprecated and `Native#load` methods were introduced, that hold the new generic definitions. So this change is now binary compatible.

--- a/build.xml
+++ b/build.xml
@@ -402,7 +402,7 @@
         <attribute name="Bundle-RequiredExecutionEnvironment" value="J2SE-1.4"/>
         <attribute name="Bundle-Vendor" value="${vendor}"/>
         <attribute name="Bundle-ActivationPolicy" value="lazy"/>
-        <attribute name="Export-Package" value="com.sun.jna,com.sun.jna.ptr,com.sun.jna.win32"/>
+        <attribute name="Export-Package" value="com.sun.jna;version=${jna.major}.${jna.minor}.${jna.revision}, com.sun.jna.ptr;version=${jna.major}.${jna.minor}.${jna.revision}, com.sun.jna.win32;version=${jna.major}.${jna.minor}.${jna.revision}"/>
         <!-- Note that no terminal "*" is included in this list,
              which will force failure on unsupported platforms.
           -->

--- a/contrib/platform/build.xml
+++ b/contrib/platform/build.xml
@@ -124,15 +124,15 @@
             <attribute name="Require-Bundle" value="com.sun.jna;bundle-version=&quot;${osgi.version}&quot;"/>
             <attribute name="Export-Package"
                    value="
-com.sun.jna.platform,
-com.sun.jna.platform.dnd,
-com.sun.jna.platform.mac,
-com.sun.jna.platform.unix,
-com.sun.jna.platform.win32,
-com.sun.jna.platform.win32.COM,
-com.sun.jna.platform.win32.COM.tlb,
-com.sun.jna.platform.win32.COM.tlb.imp,
-com.sun.jna.platform.wince
+com.sun.jna.platform;version=${osgi.version},
+com.sun.jna.platform.dnd;version=${osgi.version},
+com.sun.jna.platform.mac;version=${osgi.version},
+com.sun.jna.platform.unix;version=${osgi.version},
+com.sun.jna.platform.win32;version=${osgi.version},
+com.sun.jna.platform.win32.COM;version=${osgi.version},
+com.sun.jna.platform.win32.COM.tlb;version=${osgi.version},
+com.sun.jna.platform.win32.COM.tlb.imp;version=${osgi.version},
+com.sun.jna.platform.wince;version=${osgi.version}
 "/>
         </manifest>
     </target>

